### PR TITLE
ch4/ofi: retry in MPIDI_OFI_do_iprobe

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
We should retry the `fi_recvmsg` if it returns -FI_EAGAIN.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

File: src/mpid/ch4/netmod/ofi/ofi_impl.h [hunk 1]
```diff
@@ -101,6 +101,22 @@ int MPIDI_OFI_handle_cq_error(int vni, int nic, ssize_t ret);
...
+#define MPIDI_OFI_CALL_RETRY_RETURN(FUNC,vci_,ret) \
+    do { \
+        int _retry = MPIR_CVAR_CH4_OFI_MAX_EAGAIN_RETRY; \
+        while (1) { \
+            ret = FUNC; \
+            if (likely(ret != -FI_EAGAIN)) { \
+                break; \
+            } \
+            _retry--; \
+            MPIR_ERR_CHKANDJUMP(_retry == 0, mpi_errno, MPIX_ERR_EAGAIN, "**eagain"); \
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_impl.h:113]: Same comment as here: https://github.com/pmodels/mpich/pull/6463#discussion_r1150962668

**Comment by raffenet** [comment]
test:mpich/ch4/ofi

**Comment by raffenet** [comment]
> When it is ready, e.g., CI passes, we will test it locally with CXI.

I think this version should be OK to test. The fixups should be squashed, but no other changes.

**Comment by tarudoodi** [comment]
I'll try to get this tested by early next week.

**Comment by tarudoodi** [comment]
Hi, my first attempt at testing this failed, it did not seem directly related to this patch though. I need to take a closer look at it.

**Review by raffenet — APPROVED** [review]

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

File: src/mpid/ch4/netmod/ofi/ofi_impl.h (hunk 1 above)
Comment (raffenet) [reply src/mpid/ch4/netmod/ofi/ofi_impl.h:113]: I just saw this comment is unresolved. I'm going to push a fixup commit that modifies the retry logic to match what is in the linked PR.

**Comment by raffenet** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by tarudoodi** [comment]
I'm retesting it now. Should have an update early next week. The latest push was just a commit squash, right?

**Comment by hzhou** [comment]
> I'm retesting it now. Should have an update early next week. The latest push was just a commit squash, right?

Yes.

**Review by raffenet — APPROVED** [review]

**Pull request merged**